contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,464 @@
+    2. [Server-Side Validation](#server-side)
+    3. [Client-Side Validation](#client-side)
+    4. [Comparison between server-side and client-side Validation](#comparison)
+    5. [Existing Instances and changing the Schema](#schema-change)
+    6. [Outlook to Status Sub-Resources](#sub-resources)
+    7. [Outlook Admission Webhook](#webhooks)
+5. [Implementation Plan](#implementation)
+6. [Appendix](#appendix)
+    1. [Expressiveness of JSON-Schema](#expressiveness)
+    2. [JSON-Schema Validation Runtime Complexity](#complexity)
+    3. [Alternatives](#alternatives)
+        1. [Direct Embedding of the Schema into the Spec](#direct-embed)
+        2. [External CustomResourceSchema Type](#external-type)
+
+## <a name="overview"></a>Overview
+
+This document proposes the design and describes a way to add JSON-Schema based validation for Custom Resources.
+
+## <a name="background"></a>Background
+
+ThirdPartyResource (TPR) is deprecated and CustomResourceDefinition (CRD) is the successor which solves the fundamental issues of TPRs to form a stable base for further features.
+Currently we do not provide validation for CustomResources (CR), i.e. the CR payload is free-form JSON. However, one of the most requested [[1](https://github.com/kubernetes/features/issues/95#issuecomment-296416969)][[2](https://github.com/kubernetes/features/issues/95#issuecomment-298791881)] features is validation and this proposal seeks to add it.
+
+## <a name="goals"></a>Goals
+
+1. To provide validation for CustomResources using a declarative specification language for JSON data.
+2. To keep open the door to add other validation mechanisms later.<sup id="f2">[2](#footnote2)</sup>
+
+3. To allow server-side validation.
+4. To integrate into the existing client-side validation of kubectl.
+
+## <a name="non-goals"></a>Non-Goals
+
+1. The JSON-Schema specs can be used for creating OpenAPI documentation for CRs. The format is compatible but we won’t propose an implementation for that.
+2. A turing-complete specification language is not proposed. Instead a declarative way is proposed to express 90% of validations.
+3. For now, CRD only allows 1 version at a time. Supporting multiple versions of CRD and/or conversion of CRD is not within the scope of this proposal.
+
+## <a name="extension"></a>Proposed Extension of CustomResourceDefinition
+
+We propose to add a field “validation” to the spec of a CustomResourceDefinition. As a first validation format we propose to use [JSON-Schema](http://json-schema.org/) under `CRD.Spec.Validation.JSONSchema`. JSON-Schema is a standardized declarative specification language. Different keywords may be utilized to put constraints on the data. Thus it provides ways to make assertions about what a valid document must look like. It is already used in Swagger/OpenAPI specs in Kubernetes and hence such a CRD specification integrates cleanly into the existing infrastructure of the API server which serves these specifications,
```

Comment:
we can link to the ietf standard https://tools.ietf.org/html/draft-zyp-json-schema-04